### PR TITLE
Add local snapshots on outposts

### PR DESCRIPTION
**Pull request by ElijahQuinones**
**Is this a bug fix or adding new feature?**

fixes https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/2087

**What is this PR about? / Why do we need it?**

This PR will allow customers to pass the outpost ARN as a parameter to the VSC.
This will enable customers to have local snapshots saved onto S3 on their outpost and not just in S3 on the region their outpost is in.

**What testing is done?** 

This is currently being tested in outpost.

### CODE REVIEW

**Comment by ElijahQuinones** [comment]
/do-not-merge/work-in-progress

File: pkg/cloud/cloud.go
Comment (rdpsin) [inline pkg/cloud/cloud.go:1292]: Can you check this? I'm assuming in the case when users don't provide `outpostArn`, we end up setting this to `""` instead of `null` so `CreateSnapshot` would fail.

File: pkg/driver/constants.go
Comment (rdpsin) [inline pkg/driver/constants.go:107]: np: `OutpostArnKey` and `"outpostArn"`

File: pkg/cloud/cloud.go
Comment (ElijahQuinones) [reply pkg/cloud/cloud.go:1292]: In the case there is no outpostArn parameter in the VSC than the case in the switch statement in controller.go is never hit so the outpost arn passed to cloud.go is the empty string "" in this case we do not want to give the ec2 api an outpostarn so we drop to the else. Ether way we are creating a createSnapshotInput and passing it to ec2.

File: pkg/driver/constants.go
Comment (ElijahQuinones) [reply pkg/driver/constants.go:107]: Will change in next rev makes it much more clear

File: pkg/cloud/cloud.go
Comment (rdpsin) [reply pkg/cloud/cloud.go:1292]: Ahh, my bad. In that case, can we just create a single request object, and do the check later?

```
req := &ec2.CreateSnapshotInput {
   ...
}

if options.OutpostArn != "" {
       req.OutpostArn = aws.String(options.OutpostArn)
}
```

Comment (ElijahQuinones) [reply pkg/cloud/cloud.go:1292]: Ahhh yes we can that. Will add to next rev thank you !

File: pkg/driver/constants.go
Comment (ElijahQuinones) [reply pkg/driver/constants.go:107]: Upon reviewing the code again outpostArn will not work as we are checking for all lower case values but I will change the key in next rev.

File: pkg/driver/controller.go
Comment (rdpsin) [inline pkg/driver/controller.go:714]: Just a thought: should we validate that this is actually an ARN here? The request will fail either way when it reaches EC2, but we can avoid blasting unnecessary requests to EC2.

**Review by rdpsin — COMMENTED** [review]
Can we also add this option to the docs? I don't see any VSC parameters doc, maybe this a good opportunity to create one.

**Comment by ElijahQuinones** [comment]
Tested In outpost and confirmed we are now able to have local snapshots in outpost.

```
Type    Reason            Age   From                 Message
  ----    ------            ----  ----                 -------
  Normal  CreatingSnapshot  11m   snapshot-controller  Waiting for a snapshot default/ebs-volume-snapshot to be created by the CSI driver.
  Normal  SnapshotCreated   11m   snapshot-controller  Snapshot default/ebs-volume-snapshot was successfully created by the CSI driver.
  Normal  SnapshotReady     9m9s  snapshot-controller  Snapshot default/ebs-volume-snapshot is ready to use.
```
For testing I used the following VSC

```
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotClass
metadata:
  name: csi-aws-vsc
driver: ebs.csi.aws.com
deletionPolicy: Delete
parameters:
  outpostarn: {arn of our outpost}
```

Screenshot 2024-08-29 at 3 38 57 PM

File: pkg/driver/controller.go
Comment (ElijahQuinones) [reply pkg/driver/controller.go:714]: In my opinion I don't see a situation where someone would use anything but their outpost arn as part of the VSC if they are using this param. Validating if this is actually an arn here to me feel like bloat but I am open minded and curious what others think.

Comment (rdpsin) [reply pkg/driver/controller.go:714]: > I don't see a situation where someone would use anything but their outpost arn

This is hard to predict. :) 

Nevertheless, I'll defer this to the other team members.

Comment (torredil) [reply pkg/driver/controller.go:714]: > Just a thought: should we validate that this is actually an ARN here?

+1, I think we should consider making use of validation that the SDK provides, see [func IsARN(arn string) bool](https://github.com/aws/aws-sdk-go-v2/blob/4ed838eab2a963cb16301501c8b8c3e29dac4c20/aws/arn/arn.go#L80).

Comment (ElijahQuinones) [reply pkg/driver/controller.go:714]: sounds good will do thank you guys !

File: pkg/cloud/cloud_test.go
Comment (AndrewSirenko) [inline pkg/cloud/cloud_test.go:2062]: Once you add in the suggested `isARN` validation, consider adding an extra unit test for it.

Otherwise lgtm, thanks!

**Review by ConnorJC3 — COMMENTED** [review]
lgtm, would like to test this manually before I approve, but no issues spotted reading the code

File: pkg/cloud/cloud_test.go
Comment (ElijahQuinones) [reply pkg/cloud/cloud_test.go:2062]: sounds good thank you !

File: pkg/driver/controller.go
Comment (ElijahQuinones) [inline pkg/driver/controller.go:719]: removing now

File: docs/snapshot.md
Comment (rdpsin) [inline docs/snapshot.md:4]: np: `AvailabilityZones`

File: docs/snapshot.md
Comment (rdpsin) [inline docs/snapshot.md:53]: np: Suggest removing this. This is reiterating what's already in the doc.

Comment (ElijahQuinones) [reply docs/snapshot.md:53]: We do the same for fast snapshot. The # Fast Snapshot Restores section in the first paragraph of the linked doc. Would you suggest we remove this from both places and just have a link to the doc ?

File: pkg/driver/controller.go
Comment (AndrewSirenko) [inline pkg/driver/controller.go:713]: Nit: Would the validateCreateSnapshotRequest helper function be a better place for this arn check? (Due to having one place in the code where most of the InvalidArgument errors are created) 

The function definition is on line 790

Comment (ConnorJC3) [reply pkg/driver/controller.go:713]: In my opinion errors from parameters should be in the switch statement, that is consistent with the code and reduces duplicate parsing.

File: docs/snapshot.md
Comment (ConnorJC3) [reply docs/snapshot.md:53]: Agree - if there is information verbatim copied from the EBS docs, it should be removed (and replaced with a link).

File: pkg/driver/controller.go
Comment (AndrewSirenko) [reply pkg/driver/controller.go:713]: Where else in the switch statement are errors returned? I don't see any such code consistency. 

Seems to me that switch statement is for parsing, not for validation. 

And wouldn't we want to not make a needless call to describeSnapshots (I understand that we batch, but still).

Comment (AndrewSirenko) [reply pkg/driver/controller.go:713]: > In my opinion errors from parameters should be in the switch statement, that is consistent with the code and

The only error from parameters that is in the switch statement is the error one receives for using a non-accepted parameter. I think it's cleaner to have the function named `validateCreateSnapshotRequest` validate as much of the request as possible before any AWS API calls are wastefully made (e.g. DescribeSnapshots), than to have yet another place an error can be returned. 

But given that this is a nit, I'm fine with this conversation being resolved.

**Review by ConnorJC3 — APPROVED** [review]
Needs squash otherwise lgtm

Tested and confirmed I can create a snapshot bound to a specific outpost:
image

**Pull request merged**
